hack/netpol/pkg/main/main.go

```diff
@@ -946,3 +966,119 @@ func testPodLabelWhitelistingFromBToA() []*TestStep {
 		},
 	}
 }
+
+// This covers tests with two ingress policies with CIDRs.
+// Policy-1 with FromCIDR and ExceptCIDR and Policy-2 with FromCIDR that is same as policy-1 ExceptCIDR
+// 1) Enforce policy-1 by checking no pod is able to communicate pod from Except block
+// 2) Enforce policy-1 and policy-2 by checking all pods are able to communicate the pod on Except block in policy-1
+// For simplicity picking FromCIDR and ExceptCIDR to be same, and this is the podCIDR of node in which pod "y/a" is present.
+// TO DO: Remove this and add service deployment framework with clusterIPs to test CIDRs. Add the truth table for CIDRs.
+func testIngressOverlapCIDRBlocks() []*TestStep {
+	exceptBuilder1 := make([]string, 1)
+	exceptBuilder3 := make([]string, 1)
+
+	// It does not matter if podCIDR of pod "x/a" and "y/a" are same
+	// Build simple allowCIDR block for testing purposes.
+	allowCIDRStr := podIPs["y/a"] + "/16"
+	_, allowCIDR, err := net.ParseCIDR(allowCIDRStr)
+	if err != nil {
+		log.Errorf("Unable to parse CIDR string %s", allowCIDRStr)
+	}
+	allowCIDRStr = allowCIDR.String()
+
+	exceptBuilder1[0] = podIPs["y/a"] + "/24"
+	var exceptCIDR *net.IPNet
+	_, exceptCIDR, err = net.ParseCIDR(exceptBuilder1[0])
+	if err != nil {
+		log.Errorf("Unable to parse CIDR string %s", exceptBuilder1[0])
+	}
+	exceptBuilder1[0] = exceptCIDR.String()
+
+	// Build exceptCIDR to block just the pod "y/a"
+	exceptBuilder3[0] = podIPs["y/a"] + "/32"
+
+	allowWithExcept := func() *Reachability {
+		r := NewReachability(allPods, true)
+		r.ExpectAllIngress(Pod("x/a"), false)
+		r.Expect(Pod("x/a"), Pod("x/a"), true)
+		// Adding connectivity for all the pods in allowCIDR - exceptCIDR
+		for eachPod := range podIPs {
+			if allowCIDR.Contains(net.ParseIP(podIPs[eachPod])) && !exceptCIDR.Contains(net.ParseIP(podIPs[eachPod])) {
```

Comment:
In the example you provided, where all pods are in /24 or no pod is present in /24 along with Pod "y/a" should be ok, right? In the first case, no pod can communicate with pod "x/a" with policy-1 and all pods communicate with "x/a" with policy-1 and policy-2. With policy-1 and policy-3, all pods communicate with "x/a" except for "y/a". Can you please elaborate on the issue you see with this scenario?  

I was not completely clear about Cluster IP DNAT scenario and ingress/egress tables in OVS pipeline, but did not think elaborately. Thanks for describing the issue that it can present depending on CNI implementation.

Reply:
> In the example you provided, where all pods are in /24 or no pod is present in /24 along with Pod "y/a" should be ok, right? In the first case, no pod can communicate with pod "x/a" with policy-1 and all pods communicate with "x/a" with policy-1 and policy-2. With policy-1 and policy-3, all pods communicate with "x/a" except for "y/a". Can you please elaborate on the issue you see with this scenario?

My issue was not that the test code was incorrect, but that the test is not really deterministic: we are not necessarily testing all possible scenarios based on Pod scheduling / IP address allocation. Consider your first policy (policy1) and let me know if I'm missing something: if all the Pods are scheduled on the same Node (maybe I'm running these tests on a cluster with a single worker Node) then all the IP addresses are in the same /24, which correspond to the `except` block of your policy. In this case we are not really testing the "allow CIDR" at all. The connectivity matrix will say "false" for every `(x/a, targetPod)` tuple, except of course for `(x/a, x/a)`. Ideally we would ensure that at least one Pod always falls in the "allow CIDR" and at least one Pod always falls in the "except CIDR". Of course we can use a /32 as our "except CIDR" here, but then we still have the same problem for policy3 I think. Does that make sense or am I missing something?

But let's not block this PR for this. I can't think of a simple alternative solution right now.

Reply:
I was looking at the test as if overlap allow in policy-2 and overlap except in policy-3 is being enforced or not, when policy-1 with allow and except is there. Agree with you that if we try to enforce standalone policy-1 one.. this may not work. Thanks for the comments.